test/gtest/ucp/test_ucp_proto_select.cc

```diff
@@ -0,0 +1,447 @@
+        check_proto_select(e.worker(), config->proto_select, data, key);
+    }
+
+    static ucp_proto_select_key_t any_key()
+    {
+        ucp_proto_select_key_t key;
+        key.u64 = UINT64_MAX;
+        return key;
+    }
+
+    void connect()
+    {
+        ucp_test::init();
+        sender().connect(&receiver(), get_ep_params());
+        for (int i = 0; i < 10; ++i) {
+            progress();
+        }
+
+        EXPECT_EQ(1, sender().worker()->ep_config.length);
+        EXPECT_EQ(1, sender().worker()->rkey_config_count);
+        EXPECT_EQ(1, receiver().worker()->ep_config.length);
+        EXPECT_EQ(1, receiver().worker()->rkey_config_count);
+    }
+
+protected:
+    static size_t string_to_memunits(const std::string &str)
+    {
+        size_t range;
+        EXPECT_UCS_OK(ucs_str_to_memunits(str.c_str(), &range));
+        return range;
+    }
+
+    static bool select_elem_match(ucp_worker_h worker,
+                                  const ucp_proto_select_elem_t &select_elem,
+                                  const proto_select_data &data,
+                                  size_t range_start)
+    {
+        ucp_proto_query_attr_t proto_attr;
+        int valid = ucp_proto_select_elem_query(worker, &select_elem,
+                                                range_start, &proto_attr);
+        if (!valid) {
+            return false;
+        }
+
+        size_t range_end = string_to_memunits(data.range_end);
+        return (range_end == proto_attr.max_msg_length) &&
+               (data.desc == proto_attr.desc) &&
+               (data.config == proto_attr.config);
+    }
+
+    static void dump_select_elem(ucp_worker_h worker,
+                                 const ucp_proto_select_elem_t &select_elem)
+    {
+        size_t range_end = -1;
+        size_t range_start;
+        do {
+            range_start = range_end + 1;
+
+            ucp_proto_query_attr_t proto_attr;
+            int valid = ucp_proto_select_elem_query(worker, &select_elem,
+                                                    range_start, &proto_attr);
+            range_end = proto_attr.max_msg_length;
+            if (valid) {
+            char buf[1024];
+                sprintf(buf, "  %zu-%zu desc: %s, config: %s", range_start,
+                range_end, proto_attr.desc, proto_attr.config);
+                UCS_TEST_MESSAGE << buf;
+            }
+        } while (range_end != SIZE_MAX);
+    }
+
+    static void
+    check_proto_select_elem(ucp_worker_h worker,
+                            const ucp_proto_select_elem_t &select_elem,
+                            const proto_select_data_vec_t &data)
+    {
+        dump_select_elem(worker, select_elem);
+        for (auto &it : data) {
+            size_t start = string_to_memunits(it.range_start);
+            EXPECT_TRUE(select_elem_match(worker, select_elem, it, start));
+            /* As we cannot get range_start directly, we assert that protocol
```

Comment:
Why we cannot get range_start directly?

Reply:
Because this field does not exist neither in `ucp_proto_query_attr_t` nor in `ucp_proto_threshold_elem_t`. We kind of keep only the upper limit, and lower limit we deduce from the array: e.g. check `ucp_proto_select_elem_info`, `ucp_proto_select_write_info`

Reply:
As I can see, `select_elem_match` for `start - 1` would always return `false` since we compare `data.range_end` for current proto with `proto_attr.max_msg_length` which obviously would be different at `start - 1` point. So even if there are the same `desc` and `config` it would return false. AFAIU it would never happen now since `select_elem::thresholds` elements are merged together if same proto is selected, but I am not sure whether it is a good idea to rely on that.